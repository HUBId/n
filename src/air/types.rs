//! Core type definitions for the AIR layer.
//!
//! The final API surface is still under construction; concrete data
//! structures will be introduced as the proof system stabilises.

use core::fmt;

use crate::field::prime_field::CanonicalSerialize;
use crate::field::FieldElement as Felt;
use crate::utils::serialization::FieldElementBytes;

/// Column identifier newtype used when reporting errors.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct ColIx(usize);

impl ColIx {
    /// Creates a new column index wrapper.
    pub const fn new(ix: usize) -> Self {
        Self(ix)
    }

    /// Returns the underlying index as `usize`.
    pub const fn as_usize(self) -> usize {
        self.0
    }
}

impl fmt::Display for ColIx {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "#{}", self.0)
    }
}

/// Describes the role a trace column plays in the execution trace.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TraceRole {
    /// Main trace columns feed directly into constraint evaluation.
    Main,
    /// Auxiliary columns are derived from witness relations.
    Auxiliary,
}

impl fmt::Display for TraceRole {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            TraceRole::Main => f.write_str("main"),
            TraceRole::Auxiliary => f.write_str("auxiliary"),
        }
    }
}

/// Little-endian domain extension order.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LdeOrder {
    blowup: usize,
}

impl LdeOrder {
    /// Creates a new LDE order descriptor.
    pub fn new(blowup: usize) -> Result<Self, AirError> {
        if blowup == 0 {
            return Err(AirError::LayoutViolation(
                "lde blowup factor must be strictly positive",
            ));
        }
        if !blowup.is_power_of_two() {
            return Err(AirError::LayoutViolation(
                "lde blowup factor must be a power of two",
            ));
        }
        Ok(Self { blowup })
    }

    /// Returns the configured blowup factor.
    pub const fn blowup(&self) -> usize {
        self.blowup
    }

    fn validate_rows(&self, rows: usize) -> Result<(), AirError> {
        if rows == 0 {
            return Err(AirError::LayoutViolation(
                "execution trace must contain at least one row",
            ));
        }
        if rows % self.blowup != 0 {
            return Err(AirError::LayoutViolation(
                "lde blowup factor must divide the execution trace length",
            ));
        }
        Ok(())
    }
}

/// Degree bounds for trace and auxiliary polynomials.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct DegreeBounds {
    main: usize,
    auxiliary: usize,
}

impl DegreeBounds {
    /// Constructs new degree bounds, enforcing consistency between roles.
    pub fn new(main: usize, auxiliary: usize) -> Result<Self, AirError> {
        if main == 0 {
            return Err(AirError::LayoutViolation(
                "main trace degree bound must be positive",
            ));
        }
        if auxiliary < main {
            return Err(AirError::LayoutViolation(
                "auxiliary degree bound must be at least the main bound",
            ));
        }
        Ok(Self { main, auxiliary })
    }

    fn check_role(&self, role: TraceRole, observed_degree: usize) -> Result<(), AirError> {
        let bound = match role {
            TraceRole::Main => self.main,
            TraceRole::Auxiliary => self.auxiliary,
        };
        if observed_degree > bound {
            Err(AirError::DegreeOverflow {
                role,
                observed: observed_degree,
                allowed: bound,
            })
        } else {
            Ok(())
        }
    }

    /// Ensures the observed degree stays within the configured bounds for `role`.
    pub fn ensure(&self, role: TraceRole, observed_degree: usize) -> Result<(), AirError> {
        self.check_role(role, observed_degree)
    }
}

/// Boundary position within the execution trace.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BoundaryAt {
    /// First row of the trace.
    First,
    /// Last row of the trace.
    Last,
    /// Arbitrary row index.
    Row(usize),
}

impl fmt::Display for BoundaryAt {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            BoundaryAt::First => f.write_str("first"),
            BoundaryAt::Last => f.write_str("last"),
            BoundaryAt::Row(ix) => write!(f, "row {ix}"),
        }
    }
}

/// Metadata describing a single trace column.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceColMeta {
    /// Human readable column label.
    pub label: String,
    /// The role played by this column.
    pub role: TraceRole,
    /// Boundary values that are exposed publicly or enforced statically.
    pub boundaries: Vec<BoundaryAt>,
}

impl TraceColMeta {
    /// Creates a new trace column metadata descriptor.
    pub fn new(label: impl Into<String>, role: TraceRole) -> Self {
        Self {
            label: label.into(),
            role,
            boundaries: Vec::new(),
        }
    }

    /// Registers an additional boundary location.
    pub fn with_boundary(mut self, boundary: BoundaryAt) -> Self {
        if !self.boundaries.contains(&boundary) {
            self.boundaries.push(boundary);
        }
        self
    }

    fn allows(&self, boundary: &BoundaryAt) -> bool {
        self.boundaries.contains(boundary)
    }
}

/// Public input specification derived from the trace schema.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PublicSpec {
    /// Fields exposed to the verifier.
    pub fields: Vec<PublicFieldMeta>,
}

impl PublicSpec {
    /// Validates the specification against a trace schema.
    pub fn validate(&self, schema: &TraceSchema) -> Result<(), AirError> {
        for field in &self.fields {
            if let PublicFieldType::TraceValue { column, boundary } = &field.field_type {
                let meta = schema
                    .column_meta(*column)
                    .ok_or_else(|| AirError::SchemaMismatch {
                        what: "trace column count",
                        expected: schema.columns.len(),
                        actual: column.as_usize() + 1,
                    })?;
                if !meta.allows(boundary) {
                    return Err(AirError::BoundaryViolation {
                        column: *column,
                        boundary: *boundary,
                        detail: "boundary not declared in schema",
                    });
                }
            }
        }
        Ok(())
    }
}

/// Metadata for a single public field.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PublicFieldMeta {
    /// Identifier used in documentation and transcripts.
    pub label: String,
    /// Typing information for the exposed field.
    pub field_type: PublicFieldType,
}

/// Public field typing.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PublicFieldType {
    /// Constant value published independently of the trace.
    Constant(Felt),
    /// Value sourced from the execution trace.
    TraceValue { column: ColIx, boundary: BoundaryAt },
}

/// AIR level execution trace schema.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceSchema {
    /// Metadata for each execution trace column.
    pub columns: Vec<TraceColMeta>,
    /// Domain extension order.
    pub lde_order: LdeOrder,
    /// Degree bounds for trace polynomials.
    pub degree_bounds: DegreeBounds,
}

impl TraceSchema {
    /// Creates a new trace schema descriptor.
    pub fn new(
        columns: Vec<TraceColMeta>,
        lde_order: LdeOrder,
        degree_bounds: DegreeBounds,
    ) -> Result<Self, AirError> {
        if columns.is_empty() {
            return Err(AirError::SchemaMismatch {
                what: "trace column count",
                expected: 1,
                actual: 0,
            });
        }
        Ok(Self {
            columns,
            lde_order,
            degree_bounds,
        })
    }

    /// Returns metadata for the requested column index.
    pub fn column_meta(&self, ix: ColIx) -> Option<&TraceColMeta> {
        self.columns.get(ix.as_usize())
    }

    /// Validates an execution trace against the schema.
    pub fn validate_trace(&self, trace: &TraceData) -> Result<(), AirError> {
        if self.columns.len() != trace.num_columns() {
            return Err(AirError::SchemaMismatch {
                what: "trace column count",
                expected: self.columns.len(),
                actual: trace.num_columns(),
            });
        }
        self.lde_order.validate_rows(trace.num_rows())?;
        let observed_degree = trace.num_rows().saturating_sub(1);
        for (ix, meta) in self.columns.iter().enumerate() {
            self.degree_bounds.check_role(meta.role, observed_degree)?;
            for boundary in &meta.boundaries {
                trace.boundary_value(ColIx::new(ix), *boundary)?;
            }
        }
        Ok(())
    }
}

/// Execution trace container used for schema validation and serialization.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceData {
    columns: Vec<Vec<Felt>>,
    rows: usize,
}

impl TraceData {
    /// Creates a new trace data container.
    pub fn new(columns: Vec<Vec<Felt>>) -> Result<Self, AirError> {
        if columns.is_empty() {
            return Err(AirError::SchemaMismatch {
                what: "trace column count",
                expected: 1,
                actual: 0,
            });
        }
        let rows = columns[0].len();
        if rows == 0 {
            return Err(AirError::LayoutViolation(
                "execution trace must contain at least one row",
            ));
        }
        for column in &columns {
            if column.len() != rows {
                return Err(AirError::SchemaMismatch {
                    what: "trace row count",
                    expected: rows,
                    actual: column.len(),
                });
            }
        }
        Ok(Self { columns, rows })
    }

    /// Returns the number of columns.
    pub fn num_columns(&self) -> usize {
        self.columns.len()
    }

    /// Returns the number of rows.
    pub fn num_rows(&self) -> usize {
        self.rows
    }

    /// Returns a reference to a trace column.
    pub fn column(&self, ix: ColIx) -> Result<&[Felt], AirError> {
        self.columns
            .get(ix.as_usize())
            .map(|col| col.as_slice())
            .ok_or_else(|| AirError::SchemaMismatch {
                what: "trace column count",
                expected: self.columns.len(),
                actual: ix.as_usize() + 1,
            })
    }

    /// Returns the trace value at the specified boundary.
    pub fn boundary_value(&self, ix: ColIx, boundary: BoundaryAt) -> Result<Felt, AirError> {
        let column = self.column(ix)?;
        let row = match boundary {
            BoundaryAt::First => 0,
            BoundaryAt::Last => column.len().saturating_sub(1),
            BoundaryAt::Row(r) => r,
        };
        column.get(row).copied().ok_or(AirError::BoundaryViolation {
            column: ix,
            boundary,
            detail: "boundary outside trace length",
        })
    }

    /// Serialises a column into canonical little-endian field element bytes.
    pub fn column_bytes(&self, ix: ColIx) -> Result<Vec<FieldElementBytes>, AirError> {
        let column = self.column(ix)?;
        let mut bytes = Vec::with_capacity(column.len());
        for value in column {
            let mut buf = [0u8; 32];
            let le = value.to_bytes().map_err(|_| AirError::Serialization {
                kind: SerKind::Trace,
                detail: "non-canonical trace value",
            })?;
            buf[..le.len()].copy_from_slice(&le);
            bytes.push(FieldElementBytes { bytes: buf });
        }
        Ok(bytes)
    }
}

/// Serialization context kinds surfaced in [`AirError::Serialization`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SerKind {
    /// Execution trace serialisation.
    Trace,
    /// Public input serialisation.
    PublicInput,
    /// Transcript serialisation.
    Transcript,
}

impl fmt::Display for SerKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SerKind::Trace => f.write_str("trace data"),
            SerKind::PublicInput => f.write_str("public input"),
            SerKind::Transcript => f.write_str("transcript"),
        }
    }
}

/// Error enumeration covering deterministic AIR failures.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AirError {
    /// Schema layout mismatch (column count, row count, etc.).
    SchemaMismatch {
        /// Description of the component that failed validation.
        what: &'static str,
        /// Expected cardinality.
        expected: usize,
        /// Actual cardinality encountered.
        actual: usize,
    },
    /// Schema declared a boundary outside the trace layout.
    InvalidBoundary {
        /// Column where the invalid boundary was declared.
        column: ColIx,
        /// Boundary location that failed validation.
        boundary: BoundaryAt,
        /// Additional diagnostic context.
        detail: &'static str,
    },
    /// Boundary constraint violation.
    BoundaryViolation {
        /// Column where the violation occurred.
        column: ColIx,
        /// Boundary location.
        boundary: BoundaryAt,
        /// Additional diagnostic context.
        detail: &'static str,
    },
    /// Transition relation failed for the provided witness values.
    InvalidTransition {
        /// Column identifier.
        column: ColIx,
        /// Step index.
        step: usize,
        /// Value at `step`.
        current: Felt,
        /// Value at `step + 1`.
        next: Felt,
    },
    /// Degree bound exceeded for the specified role.
    DegreeOverflow {
        /// Role for which the overflow was detected.
        role: TraceRole,
        /// Observed degree.
        observed: usize,
        /// Allowed maximum degree.
        allowed: usize,
    },
    /// Serialization failure for one of the helper contexts.
    Serialization {
        /// Serialization context.
        kind: SerKind,
        /// Description of the failure.
        detail: &'static str,
    },
    /// Witness column admitted non-determinism.
    NonDeterministicWitness {
        /// Column identifier.
        column: ColIx,
        /// Step index where mismatch was detected.
        step: usize,
    },
    /// General layout violation (e.g. invalid parameters).
    LayoutViolation(&'static str),
}

impl fmt::Display for AirError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            AirError::SchemaMismatch {
                what,
                expected,
                actual,
            } => {
                write!(
                    f,
                    "schema mismatch for {what}: expected {expected}, got {actual}"
                )
            }
            AirError::InvalidBoundary {
                column,
                boundary,
                detail,
            } => write!(
                f,
                "invalid boundary at column {column} ({boundary}): {detail}"
            ),
            AirError::BoundaryViolation {
                column,
                boundary,
                detail,
            } => write!(
                f,
                "boundary violation at column {column} ({boundary}): {detail}"
            ),
            AirError::InvalidTransition {
                column,
                step,
                current,
                next,
            } => write!(
                f,
                "invalid transition at column {column} step {step}: {} -> {}",
                current.0, next.0
            ),
            AirError::DegreeOverflow {
                role,
                observed,
                allowed,
            } => write!(
                f,
                "degree overflow for {role} columns: observed {observed}, allowed {allowed}"
            ),
            AirError::Serialization { kind, detail } => {
                write!(f, "serialization error for {kind}: {detail}")
            }
            AirError::NonDeterministicWitness { column, step } => write!(
                f,
                "non-deterministic witness detected at column {column} step {step}"
            ),
            AirError::LayoutViolation(detail) => write!(f, "layout violation: {detail}"),
        }
    }
}

impl std::error::Error for AirError {}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_schema() -> Result<TraceSchema, AirError> {
        let columns = vec![
            TraceColMeta::new("main", TraceRole::Main)
                .with_boundary(BoundaryAt::First)
                .with_boundary(BoundaryAt::Last),
            TraceColMeta::new("aux", TraceRole::Auxiliary).with_boundary(BoundaryAt::Row(1)),
        ];
        let lde = LdeOrder::new(4)?;
        let degree = DegreeBounds::new(3, 4)?;
        TraceSchema::new(columns, lde, degree)
    }

    fn sample_trace() -> Result<TraceData, AirError> {
        TraceData::new(vec![
            vec![
                Felt::from(1u64),
                Felt::from(2u64),
                Felt::from(3u64),
                Felt::from(4u64),
            ],
            vec![
                Felt::from(5u64),
                Felt::from(6u64),
                Felt::from(7u64),
                Felt::from(8u64),
            ],
        ])
    }

    #[test]
    fn schema_validation_accepts_matching_trace() -> Result<(), AirError> {
        let schema = sample_schema()?;
        let trace = sample_trace()?;
        schema.validate_trace(&trace)?;
        Ok(())
    }

    #[test]
    fn schema_validation_detects_mismatched_columns() -> Result<(), AirError> {
        let schema = sample_schema()?;
        let trace = TraceData::new(vec![vec![
            Felt::from(1u64),
            Felt::from(2u64),
            Felt::from(3u64),
            Felt::from(4u64),
        ]])?;
        let result = schema.validate_trace(&trace);
        match result {
            Err(AirError::SchemaMismatch {
                what,
                expected,
                actual,
            }) => {
                if what != "trace column count" {
                    return Err(AirError::LayoutViolation("unexpected mismatch kind"));
                }
                if expected != 2 || actual != 1 {
                    return Err(AirError::LayoutViolation("unexpected mismatch values"));
                }
                Ok(())
            }
            Err(err) => Err(err),
            Ok(_) => Err(AirError::LayoutViolation("schema should reject trace")),
        }
    }

    #[test]
    fn trace_serialization_uses_little_endian_helpers() -> Result<(), AirError> {
        let trace = TraceData::new(vec![
            vec![Felt::from(1u64), Felt::from(0x0102_0304_0506_0708u64)],
            vec![Felt::from(9u64), Felt::from(10u64)],
        ])?;
        let column_bytes = trace.column_bytes(ColIx::new(0))?;
        if column_bytes.len() != 2 {
            return Err(AirError::LayoutViolation(
                "unexpected column serialization length",
            ));
        }
        if column_bytes[0].bytes[..8] != 1u64.to_le_bytes() {
            return Err(AirError::LayoutViolation("first element not little-endian"));
        }
        if column_bytes[0].bytes[8..].iter().any(|b| *b != 0) {
            return Err(AirError::LayoutViolation("first element padding not zero"));
        }
        if column_bytes[1].bytes[..8] != 0x0102_0304_0506_0708u64.to_le_bytes() {
            return Err(AirError::LayoutViolation(
                "second element not little-endian",
            ));
        }
        Ok(())
    }

    #[test]
    fn air_error_formatting_is_human_readable() -> Result<(), AirError> {
        let transition = AirError::InvalidTransition {
            column: ColIx::new(1),
            step: 3,
            current: Felt::from(5u64),
            next: Felt::from(7u64),
        };
        if format!("{transition}") != "invalid transition at column #1 step 3: 5 -> 7" {
            return Err(AirError::LayoutViolation(
                "unexpected transition formatting",
            ));
        }

        let serialization = AirError::Serialization {
            kind: SerKind::Trace,
            detail: "io failure",
        };
        if format!("{serialization}") != "serialization error for trace data: io failure" {
            return Err(AirError::LayoutViolation(
                "unexpected serialization formatting",
            ));
        }
        Ok(())
    }
}
