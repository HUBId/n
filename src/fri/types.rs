use core::fmt;

use crate::field::prime_field::FieldConstraintError;
use crate::hash::deterministic::DeterministicHashError;
use crate::hash::merkle::MerkleError;
use crate::params::StarkParams;

/// Transcript seed used when instantiating the FRI prover and verifier.
pub type FriTranscriptSeed = [u8; 32];

/// Security profiles supported by the FRI engine.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FriSecurityLevel {
    /// Standard profile with 64 queries.
    Standard,
    /// High security profile with 96 queries.
    HiSec,
    /// Throughput oriented profile with 48 queries.
    Throughput,
}

impl FriSecurityLevel {
    /// Returns the query budget associated with the profile.
    pub const fn query_budget(self) -> usize {
        match self {
            FriSecurityLevel::Standard => 64,
            FriSecurityLevel::HiSec => 96,
            FriSecurityLevel::Throughput => 48,
        }
    }

    /// Infers the security profile corresponding to a query budget.
    pub fn from_query_count(count: usize) -> Option<Self> {
        match count {
            c if c == FriSecurityLevel::Standard.query_budget() => Some(FriSecurityLevel::Standard),
            c if c == FriSecurityLevel::HiSec.query_budget() => Some(FriSecurityLevel::HiSec),
            c if c == FriSecurityLevel::Throughput.query_budget() => {
                Some(FriSecurityLevel::Throughput)
            }
            _ => None,
        }
    }

    pub(crate) fn tag(self) -> &'static str {
        match self {
            FriSecurityLevel::Standard => "STD",
            FriSecurityLevel::HiSec => "HISEC",
            FriSecurityLevel::Throughput => "THR",
        }
    }

    pub(crate) fn code(self) -> u8 {
        match self {
            FriSecurityLevel::Standard => 0,
            FriSecurityLevel::HiSec => 1,
            FriSecurityLevel::Throughput => 2,
        }
    }

    pub(crate) fn from_code(code: u8) -> Option<Self> {
        match code {
            0 => Some(FriSecurityLevel::Standard),
            1 => Some(FriSecurityLevel::HiSec),
            2 => Some(FriSecurityLevel::Throughput),
            _ => None,
        }
    }
}

/// Version tag attached to FRI proofs.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FriProofVersion {
    /// Initial version of the binary FRI proof format.
    V1,
}

impl FriProofVersion {
    /// Latest supported proof version.
    pub const CURRENT: FriProofVersion = FriProofVersion::V1;

    pub(crate) fn to_u16(self) -> u16 {
        match self {
            FriProofVersion::V1 => 1,
        }
    }

    pub(crate) fn from_u16(value: u16) -> Option<Self> {
        match value {
            1 => Some(FriProofVersion::V1),
            _ => None,
        }
    }
}

/// Kind markers used when reporting serialization errors.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SerKind {
    /// Parameters payload framing.
    Params,
    /// Proof body framing.
    Proof,
    /// Query openings payload framing.
    Query,
    /// Layer openings payload framing.
    Layer,
    /// Final polynomial encoding.
    FinalPolynomial,
}

impl fmt::Display for SerKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SerKind::Params => write!(f, "parameters"),
            SerKind::Proof => write!(f, "proof"),
            SerKind::Query => write!(f, "query"),
            SerKind::Layer => write!(f, "layer"),
            SerKind::FinalPolynomial => write!(f, "final polynomial"),
        }
    }
}

/// FRI verification errors mapped to the specification failure classes.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FriError {
    /// Proof version mismatch.
    VersionMismatch {
        /// Expected proof version for the verifier.
        expected: FriProofVersion,
        /// Version declared by the proof being processed.
        actual: FriProofVersion,
    },
    /// Folding constraints were violated during verification.
    FoldingConstraintViolated {
        /// Layer index where the folding constraint failed.
        layer: usize,
    },
    /// Out-of-domain sampling failed.
    OodsInvalid,
    /// Serialization failure when decoding proof artefacts.
    Serialization(SerKind),
    /// No evaluations were provided to the prover.
    EmptyCodeword,
    /// Query position exceeded the LDE domain size.
    QueryOutOfRange {
        /// Position outside the evaluation domain.
        position: usize,
    },
    /// Merkle path was malformed (index byte mismatch or inconsistent height).
    PathInvalid {
        /// Layer where the invalid path was detected.
        layer: usize,
        /// Underlying Merkle verification error.
        reason: MerkleError,
    },
    /// Merkle layer root mismatch.
    LayerRootMismatch {
        /// Layer index that produced a different root.
        layer: usize,
    },
    /// Proof declared a different security profile.
    SecurityLevelMismatch,
    /// Proof declared an unexpected number of queries.
    QueryBudgetMismatch {
        /// Expected number of queries for the security profile.
        expected: usize,
        /// Actual number of queries provided by the proof.
        actual: usize,
    },
    /// Generic structure error (missing layer, inconsistent lengths, etc.).
    InvalidStructure(&'static str),
    /// Deterministic hashing helper failed while sampling challenges.
    DeterministicHash(DeterministicHashError),
    /// Field element violated canonical encoding constraints.
    FieldConstraint(FieldConstraintError),
}

impl fmt::Display for FriError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            FriError::VersionMismatch { expected, actual } => {
                write!(
                    f,
                    "proof version mismatch (expected {expected:?}, got {actual:?})"
                )
            }
            FriError::FoldingConstraintViolated { layer } => {
                write!(f, "folding constraint violated at layer {layer}")
            }
            FriError::OodsInvalid => write!(f, "out-of-domain sampling constraint failed"),
            FriError::Serialization(kind) => {
                write!(f, "failed to serialize/deserialize FRI {kind}")
            }
            FriError::EmptyCodeword => write!(f, "codeword is empty"),
            FriError::QueryOutOfRange { position } => {
                write!(f, "query position {position} outside evaluation domain")
            }
            FriError::PathInvalid { layer, reason } => {
                write!(f, "invalid Merkle path at layer {layer}: {reason}")
            }
            FriError::LayerRootMismatch { layer } => {
                write!(f, "layer {layer} root mismatch")
            }
            FriError::SecurityLevelMismatch => write!(f, "security profile mismatch"),
            FriError::QueryBudgetMismatch { expected, actual } => write!(
                f,
                "query budget mismatch (expected {expected}, got {actual})"
            ),
            FriError::InvalidStructure(reason) => write!(f, "invalid proof structure: {reason}"),
            FriError::DeterministicHash(err) => {
                write!(f, "deterministic hash error: {err}")
            }
            FriError::FieldConstraint(err) => {
                write!(f, "field constraint violation: {err}")
            }
        }
    }
}

impl std::error::Error for FriError {}

impl From<DeterministicHashError> for FriError {
    fn from(err: DeterministicHashError) -> Self {
        FriError::DeterministicHash(err)
    }
}

impl From<FieldConstraintError> for FriError {
    fn from(err: FieldConstraintError) -> Self {
        FriError::FieldConstraint(err)
    }
}

/// Borrowed view over the parameters required when verifying a proof.
#[derive(Debug, Clone, Copy)]
pub struct FriParamsView {
    /// Declared proof version.
    pub version: FriProofVersion,
    /// Security profile used for the proof.
    pub security_level: FriSecurityLevel,
    /// Query plan identifier derived from the security level.
    pub query_plan: [u8; 32],
    /// Number of verifier queries requested by the parameter set.
    pub query_count: usize,
    /// Maximum number of folding layers executed by the prover.
    pub num_layers: usize,
    /// Log<sub>2</sub> of the initial evaluation domain size.
    pub domain_log2: usize,
}

impl FriParamsView {
    /// Constructs a view binding the static parameter fields used by the prover.
    pub fn from_params(
        params: &StarkParams,
        security_level: FriSecurityLevel,
        query_plan: [u8; 32],
    ) -> Self {
        let fri_params = params.fri();
        Self {
            version: FriProofVersion::CURRENT,
            security_level,
            query_plan,
            query_count: fri_params.queries as usize,
            num_layers: fri_params.num_layers as usize,
            domain_log2: fri_params.domain_log2 as usize,
        }
    }

    /// Returns the initial domain size derived from the parameter set.
    pub fn initial_domain_size(&self) -> usize {
        1usize << self.domain_log2
    }

    /// Returns the configured query count.
    pub fn query_count(&self) -> usize {
        self.query_count
    }

    /// Returns the number of folding layers.
    pub fn num_layers(&self) -> usize {
        self.num_layers
    }
}
