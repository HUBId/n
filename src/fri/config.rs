//! Canonical configuration descriptions for the FRI protocol.
//! Provides named profiles that are shared between prover and verifier.

/// Canonical digest representation for a FRI parameter set.
///
/// The digest is computed over the canonical serialization of the
/// folding factor, folding mode, query schedule, and target depth
/// using BLAKE3.
/// The hexadecimal strings exposed below allow integrators to pin
/// protocol parameters without embedding executable verification logic.
pub type ParameterDigest = &'static str;

/// Describes a complete FRI profile that can be referenced by the
/// STARK configuration and transcript builders.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FriProfile {
    /// Human readable profile identifier.
    pub name: &'static str,
    /// Folding factor applied to each layer; we expose the canonical binary
    /// folding schedule in this module.
    pub folding_factor: usize,
    /// Number of sampling queries performed by the verifier.
    pub query_count: usize,
    /// Target depth (logarithmic size) of the final layer after folding.
    pub target_depth: usize,
    /// Digest binding the profile to a concrete transcript layout.
    pub parameter_digest: ParameterDigest,
}

/// Standard security profile targeting typical rollup deployments.
pub const STANDARD_FRI_PROFILE: FriProfile = FriProfile {
    name: "standard",
    folding_factor: 2,
    query_count: 64,
    target_depth: 6,
    parameter_digest: "5570706d57436b4fa198d0e8c8a6d3fa01d8c348c602248b066ee8f3fe0eb566",
};

/// High security profile with an increased query budget.
pub const HISEC_FRI_PROFILE: FriProfile = FriProfile {
    name: "hisec",
    folding_factor: 2,
    query_count: 96,
    target_depth: 8,
    parameter_digest: "40b923845a41a78da7bdef951f541ae1b6b30794b0244a54b19fc5801d686b06",
};
