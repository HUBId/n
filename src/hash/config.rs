//! Hash configuration definitions shared across prover and verifier.
//! Parameters ensure deterministic hash usage within transcripts and commitments.

/// Poseidon parameter set descriptor.
#[derive(Debug, Clone)]
pub struct PoseidonParameters {
    /// Number of full rounds in the permutation.
    pub full_rounds: usize,
    /// Number of partial rounds in the permutation.
    pub partial_rounds: usize,
    /// Width of the permutation state.
    pub width: usize,
}

impl PoseidonParameters {
    /// Creates a new parameter set with deterministic defaults.
    pub fn new(full_rounds: usize, partial_rounds: usize, width: usize) -> Self {
        Self {
            full_rounds,
            partial_rounds,
            width,
        }
    }
}

/// Blake3 parameter descriptor for transcript usage.
#[derive(Debug, Clone)]
pub struct Blake3Parameters {
    /// Fixed domain separation label.
    pub label: &'static [u8],
}

impl Blake3Parameters {
    /// Creates a new descriptor.
    pub const fn new(label: &'static [u8]) -> Self {
        Self { label }
    }
}

/// Aggregate hash parameters used by the STARK engine.
#[derive(Debug, Clone)]
pub struct HashParameters {
    /// Poseidon permutation configuration.
    pub poseidon: PoseidonParameters,
    /// Blake3 transcript configuration.
    pub blake3: Blake3Parameters,
}

impl HashParameters {
    /// Constructs aggregate parameters.
    pub fn new(poseidon: PoseidonParameters, blake3: Blake3Parameters) -> Self {
        Self { poseidon, blake3 }
    }
}
