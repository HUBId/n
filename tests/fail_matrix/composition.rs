use insta::assert_debug_snapshot;
use rpp_stark::config::ProofKind as ConfigProofKind;
use rpp_stark::proof::types::VerifyError;
use rpp_stark::proof::verifier::verify_proof_bytes;

use super::{flip_composition_leaf_byte, FailMatrixFixture};

#[test]
fn composition_rejects_leaf_bytes_mismatch() {
    let fixture = FailMatrixFixture::new();
    let Some(mutated) = flip_composition_leaf_byte(&fixture.proof()) else {
        eprintln!("fixture does not expose composition openings; skipping test");
        return;
    };

    let public_inputs = fixture.public_inputs();
    let config = fixture.config();
    let context = fixture.verifier_context();

    let report = verify_proof_bytes(
        ConfigProofKind::Tx,
        &public_inputs,
        &mutated.bytes,
        &config,
        &context,
    )
    .expect("report produced");

    let error = report.error.expect("expected verification failure");
    let reason = match error {
        VerifyError::CompositionInconsistent { reason } => reason,
        other => panic!("unexpected verification outcome: {other:?}"),
    };

    let composition = mutated
        .proof
        .openings()
        .composition
        .as_ref()
        .expect("composition openings available");

    let leaves = composition.leaves.clone();
    let indices = composition.indices.clone();
    let first_index = indices
        .first()
        .copied()
        .expect("composition index available");

    let expected_reason = format!("composition_leaf_bytes_mismatch:pos=0:index={first_index}");
    assert_eq!(reason, expected_reason);

    assert_debug_snapshot!("rejects_leaf_bytes_mismatch__leaves", leaves);
    assert_debug_snapshot!("rejects_leaf_bytes_mismatch__indices", &indices);
    assert_debug_snapshot!("rejects_leaf_bytes_mismatch__reason", &reason);
}
