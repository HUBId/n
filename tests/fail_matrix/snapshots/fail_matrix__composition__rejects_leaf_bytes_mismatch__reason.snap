---
source: tests/fail_matrix/composition.rs
expression: "&reason"
---
"composition_leaf_bytes_mismatch:pos=0:index=3"
